settings: Replace extraneous pointer cursors on text with text cursor.

palashb01 opened the pull request:
- [x] Replace extraneous pointer cursors on text with text cursor.
- [x] Convert labels with input fields to focus on fields when clicked.

--------------------------
Fixes: #21769

-----------------------------------

**Screenshots and screen captures:**

|Before | After |
|----------|----------|
| ![before](https://user-images.githubusercontent.com/66828942/223761086-3c13d680-c017-41f4-bf53-711c569895f8.gif)| ![after](https://user-images.githubusercontent.com/66828942/223761177-cf3a7895-72cc-4a7b-8fe1-4418b9f80be3.gif) |
|![old-](https://user-images.githubusercontent.com/66828942/224184057-1124a66f-312a-4d62-9aa9-08fbbf520392.gif) | ![new_](https://user-images.githubusercontent.com/66828942/224184080-d2a91272-1c0f-4c8e-9a45-8870322444d8.gif) |




<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

palashb01 commented:
Hey @alya , could you please review the PR, Thanks.

alya commented:
Thanks! The screen capture looks good. Is this the only place that still needs fixing?

palashb01 commented:
> Thanks! The screen capture looks good. Is this the only place that still needs fixing?

Yes, I have checked all the sections of the settings panel and stream panel, and only these labels needed to be changed.

alya commented:
Cool, thanks! @sahil839 could you please take a look?

sahil839 commented:
Clicking on label elements can be used to focus on the input if `for` attribute is correctly set for the label element. This may not be helpful for select elements as such but still can be useful for text type inputs as user has more clickable area to focus on the input and they can then start typing after the input is focussed.

palashb01 commented:
> Clicking on label elements can be used to focus on the input if `for` attribute is correctly set for the label element. This may not be helpful for select elements as such but still can be useful for text type inputs as user has more clickable area to focus on the input and they can then start typing after the input is focussed.

Hey @sahil839 , I have made all the suggested changes and updated the PR description too, could you please review the PR, Thanks.

sahil839 commented:
Ohh I intended to have some feedback from Alya first. Sorry for the confusion. 

> Clicking on label elements can be used to focus on the input if `for` attribute is correctly set for the label element. This may not be helpful for select elements as such but still can be useful for text type inputs as user has more clickable area to focus on the input and they can then start typing after the input is focussed.

@alya any thoughts?

alya commented:
Ah, apologies, I wasn't quite clear on the question. Could you guys explain more specifically what options we're deciding between?

sahil839 commented:
Currently, for some labels clicking on them focuses the input or select element. This happens if `for` element for the label is set correctly. So, do we want to keep the `pointer` cursor since clicking actually does something. 

![label-cursor](https://user-images.githubusercontent.com/35494118/224245644-a8e5ab53-a14c-44b0-a460-38180c5682e8.gif)

palashb01 commented:
> Ah, apologies, I wasn't quite clear on the question. Could you guys explain more specifically what options we're deciding between?
> 
> 

i have attached the screenshot of the changes in the PR description, we are referring to before and after of the changes (2nd row in the table)

alya commented:
Ah, I see, thanks!

Keeping the pointer cursor when clicking on the label focuses then input seems reasonable to me, but it might be worth posting in #design for feedback. I'm not sure what will feel most natural.

palashb01 commented:
@alya, I have updated the PR as per the latest discussion on CZO [Thread](https://chat.zulip.org/#narrow/stream/101-design/topic/disabled.20checkbox.20label.20cursor), could you please review the PR? Thanks :)

alya commented:
Cool, thanks! I haven't tested everything in the settings, but everything I checked looked good to me. @sahil839 could you take another look an mark this PR for integration review if/when it's ready from your perspective?

sahil839 commented on web/templates/settings/custom_user_profile_field.hbs:
```diff
@@ -1,5 +1,5 @@
 <div class="custom_user_field" name="{{ field.name }}" data-field-id="{{ field.id }}">
-    <label class="inline-block" for="{{ field.name }}" class="title">{{ field.name }}</label>
+    <label class="inline-block text_cursor title" for="{{ field.name }}">{{ field.name }}</label>
```
The `text_cursor` should be added in next commit. This commit should only fix duplicate class attributes.

palashb01 replied:
Added a separate commit for this.

sahil839 commented on web/styles/settings.css:
```diff
@@ -1743,6 +1743,9 @@ $option_title_width: 180px;
     }
 }
 
+.text_cursor {
```
I think would be better to move this code in `app_components.css`.

palashb01 replied:
Moved it to `app_components.css`

sahil839 commented on web/templates/stream_settings/stream_types.hbs:
```diff
@@ -41,8 +41,8 @@
-    <label class="dropdown-title">{{t 'Who can unsubscribe others from this stream?'}}
-    </label>
+    <div class="dropdown-title">{{t 'Who can unsubscribe others from this stream?'}}
+    </div>
```
I am not sure whether we want to change this to `div`, since the text still acts as a label for the dropdown setting.

sahil839 replied:
I know we do not use an `input` or `select` element here, but still worth pointing out in case we want feedback from others.

palashb01 replied:
Clicking on these `label` tags doesn't focus anywhere, which is not the expected behavior of label tags. Thus, using them seems logically incorrect. I replaced them with a more appropriate element `div`. I added a separate commit with the reasoning for this and also fixed the broken colors of the help links

timabbott replied:
Why is the expected behavior of clicking on the label for a dropdown to focus something? 

(I do think it's probably a minor bug that this doesn't have a `for=` applied).

palashb01 replied:
Fixing this bug in recent commit.

sahil839 commented on web/templates/settings/notification_settings.hbs:
```diff
@@ -68,9 +68,9 @@
         {{/each}}
 
-        <label for="notification_sound">
+        <div for="notification_sound">
```
I cannot understand why this is changed to `div`, since this is associated with the `select` element.

sahil839 replied:
And anyways, I think we should have this and similar changes in a separate commit with explaining the reasoning in commit message.

palashb01 replied:
Have made a separate commit with proper reasoning for this.

sahil839 commented on web/templates/stream_settings/stream_types.hbs:
```diff
@@ -28,9 +28,9 @@
 
 <div class="input-group">
-    <label class="dropdown-title">{{t 'Who can post to the stream?'}}
+    <div class="dropdown-title">{{t 'Who can post to the stream?'}}
```
Same question as above, this is also associated with the `select` element.

sahil839 reviewed (commented):
Posted some comments. Also, I noticed that the `?` links are blue where the `label` is changed to `div`, so this should be taken care if we are doing these changes.

![Screenshot from 2023-03-30 14-49-35](https://user-images.githubusercontent.com/35494118/228791038-a641cd1d-2c96-4d14-8505-e3332b099c9e.png)

palashb01 commented:
Hey @sahil839 , I have refactored the code according to your suggestions and split the commit into subparts with proper reasoning for each change. Could you please review it again? Thanks :)

sahil839 commented:
Currently, some select elements are focussed on clicking the label and some are not, probably due to the `for` attribute is not set correctly.

Regarding changing some of the `label` elements -
I am not sure about the radio type settings, but for settings like `Notification sound` setting, which has a select element, we can have focus on the element on clicking the label. It is not being focussed currently either due to incorrect `for` attribute or due to the label element not positioned correctly.

And for the dropdown list widget settings, there is no `for` attribute set to the label element, but if set one and set it to the id of button element, then clicking on the label would open the dropdown which would make its behavior inconsistent with the select elements, so I am not sure about it. But we can probably still use `label` for it instead of `div` since they both are still visually associated.

sahil839 commented:
We can probably have more feedback from others before proceeding with the points mentioned in above comment.

palashb01 commented:
> We can probably have more feedback from others before proceeding with the points mentioned in above comment.

Sure, will start a thread for this.

timabbott commented on web/styles/settings.css:
```diff
@@ -14,6 +14,7 @@ label {
 label,
 h4,
 h3,
+div,
```
Does this add new CSS to all links inside divs in the app, aka just about every link? I don't think we want to do this.

My guess is there's a better way to do this, but also please take some time to read about thoughtful CSS scoping -- it's an important topic and it shouldn't be reviewers' job to catch this sort of issue in your work.

palashb01 replied:
Sorry for the inconvenience. I was experimenting with different things and forgot to push the correct logic for this fix. I will review the changes more carefully next time.

timabbott commented:
Thanks for working on this issue, but I really don't like the direction of this PR; it's the kind of thing that will make our settings CSS harder to maintain, not easier. Structurally, our settings UI is filled with components -- a handful of types elements: checkboxes, dropdowns, dropdown_list_widgets, and a few other things. 

We should figure out how to fix this by adjusting the common CSS for those components, if necessary including introducing a new semantic class `setting-label` or `settings-select-label` class, but not adding a new CSS class to every single label element within settings. And certainly we shouldn't change `label` elements into `div` elements. 

I would rather we do the work to make our reusable components behave how we want to by default on this front (possibly in a new PR, as this one is pretty cluttered already), and then potentially work through how to manage whatever corner case components might require some fiddling with the HTML (adding `for` attributes, etc.) .
